quickwit-actors/src/actor.rs

```diff
@@ -0,0 +1,287 @@
+use std::fmt;
+use std::ops::Deref;
+use std::time::Duration;
+use std::{any::type_name, sync::Arc};
+use thiserror::Error;
+use tracing::{debug, error};
+
+use crate::actor_handle::ActorMessage;
+use crate::mailbox::Command;
+use crate::scheduler::{Callback, SchedulerMessage};
+use crate::{
+    actor_state::{ActorState, AtomicState},
+    progress::{Progress, ProtectZoneGuard},
+    AsyncActor, KillSwitch, Mailbox, QueueCapacity, SendError, SyncActor,
+};
+
+// While the lack of message cannot pause a problem with heartbeating,  sending a message to a saturated channel
+// can be interpreted as a blocked actor.
+
+#[derive(Error, Debug)]
+pub enum ActorTermination {
+    /// The actor was stopped upon reception of a Command.
+    #[error("On Demand")]
+    OnDemand,
+    /// The actor tried to send a message to a dowstream actor and failed.
+    /// The logic ruled that the actor should be killed.
+    #[error("Downstream actor closed connection")]
+    DownstreamClosed,
+    /// Some unexpected error happened.
+    #[error("Failure")]
+    Failure(#[from] anyhow::Error),
+    /// The actor terminated, as it identified it reached a state where it
+    /// would not send any more message.
+    ///
+    /// It happens either because all of the existing mailbox where dropped and
+    /// the actor message queue was exhausted, or because the actor
+    /// returned `Err(ActorTermination::Finished)` from its `process_msg` function.
+    #[error("Finished")]
+    Finished,
+    /// The actor was terminated due to its kill switch.
+    #[error("Actor stopped working as the killswitch was pushed by another actor.")]
+    KillSwitch,
+}
+
+impl ActorTermination {
+    pub fn is_finished(&self) -> bool {
+        match self {
+            ActorTermination::Finished => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_failure(&self) -> bool {
+        match self {
+            ActorTermination::OnDemand => false,
+            ActorTermination::DownstreamClosed => true,
+            ActorTermination::Failure(_) => true,
+            ActorTermination::Finished => false,
+            ActorTermination::KillSwitch => false,
+        }
+    }
+}
+
+impl From<SendError> for ActorTermination {
+    fn from(_: SendError) -> Self {
+        ActorTermination::DownstreamClosed
+    }
+}
+
+/// An actor has an internal state and processes a stream of message.
```

Comment:
I think it would be good to the add why we have this and where it's used (maybe also in the `README.md` in the project)

e.g. these actors are used only for indexing I guess, I that case `IndexingActor` would be a better name

Reply:
Right now it is only for indexing yes. But I prefer to separate the concerns and keep `quickwit-actors` entirely
agnostic. There are no real reasons why it shouldn't be... 

And if it is not killed by Darwinism, it will very likely be reused in other parts of quickwit too.

Reply:
I definitely need to add a README.